### DISCUSSION TITLE
Pointers?

### DISCUSSION THREAD

**Discussion by TheColonel2688** [JeremyMedders/LogixLibraries#8] [General]
> Pointers vs. InOut / ByRef data handling and (safe!) examples
You mention pointers in the To-Do section of the reason for me. I am confused as Rockwell doesn't have Pointers as far as I know.

What is this referring too?

**Comment by kyle-github** [comment]
At least in languages with unconstrained pointers like C and C++, this is playing with fire.   Rockwell's programming model does not really have them.  However, you can use aliases to get some of the benefits.   That's about as close as you can get.

The InOut/ByRef probably refers to parameter handling when calling a procedure.   If you have pass by value, then the value is copied and passed to a procedure.  I can't really think of how to do that with Rockwell's model.   ByRef means that instead of passing a copy of the data, you pass a reference to it.   If it is a reference (usually a pointer under the hood) you can modify the data that is passed and the caller of the procedure will see the difference.

The calling convention of (at least in ladder logic) procedures is actually a lot closer to early FORTRAN than anything else I can think of.  E.g. No stack.  When you call a procedure, I strongly suspect that the arguments are pass ByRef.

All this is guesses based on behavior I've seen.

**Reply by TheColonel2688** [reply]
None of that answers the question of why he put creating examples about them on his to-do list.

**Reply by JeremyMedders** [reply]
These examples and their explanation/tutorial are coming with v2.

I've been grinding out features, fixes, and _**testing, testing, testing**_ for v2 like a madman lately.

An example of using a pointer to write into local AOI strings already exists in v1.

This snippet is found in a Config routine for one of the drives in my test setup:
```
_StrCond[0] := 'Device Status OK';
_StrCond[1] := 'Communication OK';
_StrCond[2] := 'Stop Input OK';
_StrCond[3] := 'Device Safety OK';
_StrCond[4] := 'Motor Breaker OK';
_StrCond[5] := 'Motor Disconnect OK';
_StrCond[6] := 'Local E-Stop OK';

COP(_StrCond[0], Interlocks.Ptr_Condition00, Interlocks.Val_StrArraySize);
```

- There's a shared temporary array for setting each string up (**_StrCond**).
- This entire array is then pushed into the Interlock backing data by using a COP() call.

**How it works**

- **_StrCond[0]** is the _base_ source element.
- **Ptr_Condition00** is the _base_ destination element.
- **Ptr_Condition00** is an AOI parameter that is aliased to the _first member of the first string_ in the Conditions array (Conditions[0].LEN).
- This makes **Ptr_Condition00** a DINT, and it appears as such both externally and, more importantly, to COP() to be such.

**So...**

- COP() is provided the base Source element (STR_40) and the base Destination element (DINT)
- It is told, in its third parameter, **_how many strides_** of the Destination element's type (DINT), to copy from Source to Destination.
- The number of strides, **_Interlocks.Val_StrArraySize_**, is calculated internally to be the number of DINT-sized blocks that compose the Interlocks.Conditions array.
- 32 x STR_40 (44 bytes per) = 1408 bytes, or 352 DINTs.
